components/proxy/src/main/java/com/hotels/styx/startup/extensions/FailureHandling.java

```diff
@@ -0,0 +1,118 @@
+/*
+  Copyright (C) 2013-2019 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.startup.extensions;
+
+import com.hotels.styx.common.FailureHandlingStrategy;
+import com.hotels.styx.common.Pair;
+import com.hotels.styx.proxy.plugin.NamedPlugin;
+import com.hotels.styx.proxy.plugin.PluginStartupException;
+import com.hotels.styx.spi.config.SpiExtension;
+import org.slf4j.Logger;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * Strategies for handling failure during start-up.
+ */
+final class FailureHandling {
+    private static final Logger LOGGER = getLogger(FailureHandling.class);
+
+    static final FailureHandlingStrategy<Pair<String, SpiExtension>, ConfiguredPluginFactory> PLUGIN_FACTORY_LOADING_FAILURE_HANDLING_STRATEGY =
+            new FailureHandlingStrategy.Builder<Pair<String, SpiExtension>, ConfiguredPluginFactory>()
+
+                    .doImmediatelyOnEachFailure((plugin, err) -> logIndividualFailure(new FailedPlugin(plugin), err))
+
+                    .doOnFailuresAfterAllProcessing(failures -> {
+                        throw new PluginStartupException(afterFailuresErrorMessage(transformKeys(failures, FailedPlugin::new)));
```

Comment:
it seems transformKeys()'s second argument is always  "FailedPlugin::new" . Then why do we need it be an argument at all?